File: src/Planet.hs
module Planet (
  Planet,
  create,
  addMarker,
  checkMarker,
  checkCoords
) where

import Data.Set (Set)
import qualified Data.Set as Set

import Bot (Coord, Heading)

data Planet = Planet {
  width :: Int,
  height :: Int,
  markers :: Set Coord
} deriving (Show, Eq)

create :: Int -> Int -> Planet
create w h = Planet { width = w, height = h, markers = Set.empty }

addMarker :: Planet -> Coord -> Planet
addMarker planet coord =
    planet { markers = Set.insert coord $ markers planet }

checkMarker :: Planet -> Coord -> Bool
checkMarker planet coord =
    Set.member coord $ markers planet

checkCoords :: Planet -> Coord -> Bool
checkCoords (Planet w h _) (x, y) = x < w && x >= 0 && y < h && y >= 0

File: scala/src/main/scala/Planet.scala
package com.rumblesan.marsrover

case class Planet(width: Int, height: Int, markers: Set[(Int, Int)])

object Planet {

  def validPosition(planet: Planet, xPos: Int, yPos: Int): Boolean = {
    (xPos >= 0 && xPos <= planet.width && yPos >= 0 && yPos <= planet.height)
  }

  def addMarker(planet: Planet, xPos: Int, yPos: Int): Planet = {
    planet.copy(markers = planet.markers + ((xPos, yPos)))
  }

  def checkMarker(planet: Planet, xPos: Int, yPos: Int): Boolean = {
    planet.markers.contains((xPos, yPos))
  }

}


